kinds.go

```diff
@@ -0,0 +1,123 @@
+	// to find the appropriate schema
+	SoftwareVersion string `json:"software"`
+
+	// Human readable descriptions of the changes in this version
+	Changelog []string `json:"changelog,omitempty"`
+
+	// The YYYY-MM-DD this version was published (or empty if not yet published)
+	Published string `json:"published,omitempty"`
+
+	// JSONSchema hash
+	Signature string `json:"signature,omitempty"` // ?? hash of the json schema
+}
+
+type MachineNames struct {
+	// This is used in k8s URLs
+	Plural string `json:"plural,omitempty"`
+
+	// Used as an alias in the display
+	Singular string `json:"singular,omitempty"`
+
+	// Optional shorter names that can be matched in a CLI
+	Short []string `json:"short,omitempty"`
+}
+
+type Kind interface {
+	// Get general information about this kind
+	GetKindInfo() KindInfo
+
+	// Get the latest version
+	CurrentVersion() string
+
+	// Get all versions
+	GetVersions() []VersionInfo
+
+	// Return a JSON schema definition for the selected version
+	// When composition is required, the slots will have an any node
+	// TODO? include an option to have `AnyOf(latest known options)`
+	GetJSONSchema(version string) (string, error)
+}
+
+// A kind that manages a k8s style resource
+type ResourceKind interface {
+	Kind
+
+	// K8S style machine names for this kind
+	GetMachineNames() MachineNames
+
+	// Read data into a Resource, when strict is true, all validation rules will be checked
+	// otherwise a resource will be created if possible, but all validation may not have been run
+	Read(reader io.Reader, strict bool) (Resource, error)
+
+	// Migrate from one object to another version
+	// NOTE, this may require database calls
+	Migrate(ctx context.Context, obj Resource, targetVersion string) (Resource, error)
+}
+
+// Function that can be used to migrate resources
+type ResourceMigrator = func(ctx context.Context, obj Resource, targetVersion string) (Resource, error)
+
+// Composable kinds are used within other resources -- they do not exist on their own
+// NOTE: this interface is still an untested sketch.  We need to implement something real
+// before settling on the real method signatures.  However, it is still useful to have the
+// common container metadata so that we can produce consistent docs and clients.
+type ComposableKind interface {
+	Kind
+
+	// eg: panel(options+fieldconfig) | transformation | dataquery | matcher
+	GetComposableType() string
+
+	// panel currently has Options + FieldConfig
+	// TODO?? can we get rid of slots and just have two composable kinds in the same plugin?
+	GetComposableSlots() []string
+
+	// Given an object (at a version) check that it is valid
+	Validate(obj any, sourceVersion string) error
+
+	// Migrate from one version of the object to another
+	Migrate(ctx context.Context, obj any, sourceVersion string, targetVersion string) (any, error)
+}
+
+// Not yet implemented, but will be required for kinds that require composition
+type KindRegistry interface {
```

Comment:
I'm a bit confused by seeing code generation being mentioned here.

Isn't this a runtime service designed to be used by services (plugins or whatever) to query available kinds? I guess for code generation we'll need something like the [`kind-registry`](https://github.com/grafana/kind-registry) ([repo](https://github.com/grafana/kind-registry)), not as a runtime service but like a "central repository".

Could you clarify that, please? @ryantxu

Reply:
For composition to work, we will *need* a runtime service saying what is known right now -- The central repository project... I am not really sure what value it has TBH, I assumed it is a codegen requirement but not really sure.

Reply:
I guess the alternative would be to provide a separate set of interfaces/abstractions that should be used to generate code/schemas.

Reply:
> For composition to work, we will need a runtime service saying what is known right now

Gotcha, thanks! We'll most likely need to clearly identify/separate both to avoid confusion from now on.

> This might be an incorrect assumption from my side but currently we are using the kindsys during code/schema- generation while producing types/schemas.

This is correct, although we'll need to update code generation pipelines sooner than later, at very least to reduce coupling between those and Thema/CUE. Looks like if we need to have a "standard schema language" for that, it's mostly likely going to be OAPI/JSON Schema.

> You are correct regarding the kind-registry but at the same time it doesn't make sense (I think) for e.g. plugins/core grafana to rely on kinds that they provide/own from the registry when they just can use the "local" ones which potentially is a version not yet being published to the registry.

Yeah, I agree @mckn that for local uses, it makes no sense. I guess that's similar for different schemas/kinds defined within the same app.

> I guess the alternative would be to provide a separate set of interfaces/abstractions that should be used to generate code/schemas.

I'm not even sure if those need to be in Kindsys, and/or if entirely there.